src/Servers/Connections.Abstractions/src/StreamContext.cs

```diff
@@ -0,0 +1,33 @@
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading;
+
+namespace Microsoft.AspNetCore.Connections
+{
+    public abstract class StreamContext : ConnectionContext
+    {
+        /// <summary>
+        /// Triggered when the client stream is closed.
+        /// </summary>
+        public virtual CancellationToken StreamClosed { get; set; }
+
+        /// <summary>
+        /// Gets or sets a unique identifier to represent this stream in trace logs.
+        /// </summary>
+        public abstract long StreamId { get; set; }
+
+        /// <summary>
+        /// Represents the direction
+        /// </summary>
+        public abstract Direction Direction { get; }
+    }
+
+    public enum Direction
+    {
+        BidirectionalInbound,
+        BidirectionalOutbound,
+        UnidirectionalInbound,
+        UnidirectionalOutbound
```

Comment:
These seem a little too... QUICy? I'm concerned about the durability of the abstraction, particularly after being burned by `ConnectionContext`. To me that presents two options:

1. Demote `Direction` from `StreamContext`. Make it something QUIC specific (`QuicStreamType`) and put it in a feature

1. Find a design for `Direction` that feels more durable. Perhaps splitting inbound/outbound from bidirectional/unidirectional? Or... splitting it in to `CanReceive` and `CanSend` (feels... awkward)?

Reply:
I couldn't figure out a great way to represent this either. Do you agree that when creating a stream, it should take a parameter specifying whether it should be unidirectional or bidirectional?

Reply:
For QUIC, sure. For an arbitrary multiplexed protocol? I'm not sure. We talked about a property bag or `IFeatureCollection` here. Perhaps that's an option.

Reply:
I already added the feature collection. I'll go ahead and remove the unidirectional bool.